service/interpreter/stateExecutionCounter.go

```diff
@@ -109,8 +109,35 @@ func (e *StateExecutionCounter) MarkStateIdExecutingIfNotYet(stateReqs []StateRe
 	}
 	e.totalCurrentlyExecutingCount += numOfNew
 
+	// Optimization: don't upsert SAs if currentSAsValues == stateIdCurrentlyExecutingCounts keys
+	if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() && needsUpdateSA {
+		sas, err := e.provider.GetSearchAttributes(e.ctx, []iwfidl.SearchAttributeKeyAndType{
+			{Key: ptr.Any(service.SearchAttributeExecutingStateIds),
+				ValueType: ptr.Any(iwfidl.KEYWORD_ARRAY)},
+		})
+		if err != nil {
+			e.provider.GetLogger(e.ctx).Error("error for GetSearchAttributes", err)
+		}
+
+		var currentSAsValues []string
+
+		currentSAs, ok := sas[service.SearchAttributeExecutingStateIds]
+		if ok {
+			currentSAsValues = currentSAs.StringArrayValue
+		}
+
+		var executingStateIds []string
+		for sid := range e.stateIdCurrentlyExecutingCounts {
+			executingStateIds = append(executingStateIds, sid)
+		}
+
+		if reflect.DeepEqual(currentSAsValues, executingStateIds) {
```

Comment:
Actually, I think we should use something like elementMatch to compare ? Because the order could be different https://stackoverflow.com/questions/36000487/check-for-equality-on-slices-without-order

Reply:
I sort the slice (line 170)

Reply:
But I agree doing elementMatch would be easier. I learned about it after I added sorting... 🙃

Reply:
Nice. Sorry I didn’t read your latest code yet. I just thinking about it and wondering why you had NDE yesterday. This could be the reasons. Especially doing it in the update/refresh method for both MarkExecuting and MarkCompleted, it increases a lot the likelihood of NDE because of the randomness

Reply:
Have you push the latest code? I don’t see the sorting ( do you mean sorting the keys of the map, right?)

Reply:
https://github.com/indeedeng/iwf/pull/460#pullrequestreview-2391203570

Reply:
Oh, I see the confusion. I was focusing on the tests... my bad. Yes, I should add sorting or elementMatch here.

Reply:
Added now

Reply:
Now with this sorting, can you move the whole optimization back to `refreshIwfExecutingStateIdSearchAttribute` again? I think if this solves the NDE, it will be also help the issue of https://github.com/indeedeng/iwf/pull/460#discussion_r1814203277